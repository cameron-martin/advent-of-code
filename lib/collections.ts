export function maxBy<T>(array: T[], by: (elem: T) => number): T {
    return array.reduce((currentHighest, elem) => by(elem) > by(currentHighest) ? elem : currentHighest);
}

export function minBy<T>(array: T[], by: (elem: T) => number): T {
    return array.reduce((currentLowest, elem) => by(elem) < by(currentLowest) ? elem : currentLowest);
}

export function sum(numbers: number[]): number {
    return numbers.reduce((a, b) => a + b, 0);
}

export function product(numbers: number[]): number {
    return numbers.reduce((a, b) => a * b, 1);
}

type Arrays<T> = {
    [K in keyof T]: Array<T[K]>;
};

export function cartesianProduct<T extends any[]>(...args:  Arrays<T>): T[] {
    if(args.length === 0) return [[]] as any;

    const [init, ...rest] = args;

    const restProduct = cartesianProduct(...rest);

    return init.flatMap(val => restProduct.map(rest => [val, ...rest]) as any);
}

export function keyBy<T>(array: T[], f: (x: T) => string): Record<string, T> {
    const result: Record<string, T> = {};

    array.forEach((item) => {
        result[f(item)] = item;
    });

    return result;
}
