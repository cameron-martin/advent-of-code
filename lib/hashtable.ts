export interface Hashable {
    readonly hash: number;
}

export interface Equatable {
    equals(other: this): boolean;
}

interface Item<K, V> {
    key: K;
    value: V;
}

export class HashTable<K extends Hashable & Equatable, V> {
    private buckets = new Map<number, Item<K, V>[]>();

    private getBucket(key: K) {
        const hash = key.hash;

        if(!this.buckets.has(hash)) this.buckets.set(hash, []);
        return this.buckets.get(hash)!;
    }

    set(key: K, value: V): void {
        let bucket = this.getBucket(key);

        const currentIndex = bucket.findIndex(item => item.key.equals(key));

        if(currentIndex === -1) {
            bucket.push({ key, value });
        } else {
            bucket.splice(currentIndex, 1, { key, value });
        }
    }

    get(key: K): V | undefined {
        const bucket = this.getBucket(key);

        const item = bucket.find(item => item.key.equals(key));

        if(item !== undefined) {
            return item.value;
        }
    }

    has(key: K): boolean {
        const bucket = this.getBucket(key);

        return bucket.some(item => item.key.equals(key));
    }
}
