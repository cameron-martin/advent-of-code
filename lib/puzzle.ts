import path from 'path';

export interface PuzzleResult {
    part1: string | number;
    part2?: string | number;
}

type PuzzleDefintion = () => PuzzleResult | Promise<PuzzleResult>;

export const puzzle = (puzzleDefintion: PuzzleDefintion): PuzzleDefintion => puzzleDefintion;

export async function getPuzzleResult(year: number, day: number): Promise<PuzzleResult> {
    const puzzle = require(`../solutions/${year}/${day}/solve`).default;

    return puzzle();
}
