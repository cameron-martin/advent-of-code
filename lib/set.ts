export function subset<T>(a: Set<T>, b: Set<T>): boolean {
    let isSubset = true;

    a.forEach((elem) => {
        if(!b.has(elem)) {
            isSubset = false;
        }
    });

    return isSubset;
}

export function intersection<T>(...sets: Set<T>[]): Set<T> {
    if(sets.length === 0) {
        throw new Error('Cannot take intersection of no sets');
    }

    const [head, ...tail] = sets;

    let result = new Set<T>();

    head.forEach((item) => {
        if(tail.every((otherSet) => otherSet.has(item))) {
            result.add(item);
        }
    });

    return result;
}

export function union<T>(...sets: Set<T>[]): Set<T> {
    const result = new Set<T>();

    sets.forEach((set) => {
        set.forEach(item => {
            result.add(item);
        });
    });

    return result;
}