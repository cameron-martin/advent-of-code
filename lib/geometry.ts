import { minBy, maxBy } from "./collections";
import { Hashable } from "./hashtable";

export class Coord implements Hashable {
    static origin = new Coord(0, 0);

    constructor(public x: number, public y: number) {}

    get hash() {
        // TODO: This has a lot of collisions - reconsider this if performance becomes an issue.
        return this.x + this.y;
    }

    equals(other: Coord) {
        return this.x === other.x && this.y === other.y;
    }

    add(other: Coord) {
        return new Coord(this.x + other.x, this.y + other.y);
    }
}

export class Box {
    static bounding(coords: Coord[]) {
        return new Box(
            new Coord(
                minBy(coords, coord => coord.x).x,
                minBy(coords, coord => coord.y).y,
            ),
            new Coord(
                maxBy(coords, coord => coord.x).x,
                maxBy(coords, coord => coord.y).y,
            )
        );
    }

    constructor(public min: Coord, public max: Coord) {}

    *cells(): IterableIterator<Coord> {
        for(let x = this.min.x; x <= this.max.x; x++) {
            for(let y = this.min.y; y <= this.max.y; y++) {
                yield new Coord(x, y);
            }
        }
    }

    isOnBoundary(coord: Coord) {
        return [this.min.x, this.max.x].includes(coord.x) || [this.min.y, this.max.y].includes(coord.y);
    }
}

export function manhattanDistance(coord1: Coord, coord2: Coord) {
    return Math.abs(coord1.x - coord2.x) + Math.abs(coord1.y - coord2.y);
}
