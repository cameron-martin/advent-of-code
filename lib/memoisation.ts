import { Hashable, Equatable, HashTable } from "./hashtable";

export function memoise<T extends Hashable & Equatable, R>(f: (arg: T) => R): (arg: T) => R {
    const cache = new HashTable<T, R>();

    return function(arg) {
        if(!cache.has(arg)) cache.set(arg, f(arg));

        return cache.get(arg)!;
    };
}
