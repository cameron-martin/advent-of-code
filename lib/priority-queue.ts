import { Equatable } from "./hashtable";

function getParent(index: number): number {
    return Math.floor((index + 1) / 2) - 1;
}

function hasParent(index: number) {
    return index !== 0;
}

const getLeftChildIndex = (index: number) => 2 * index + 1;
const getRightChildIndex = (index: number) => 2 * index + 2;

interface HeapNode<T> {
    value: T;
    priority: number; 
}

export class PriorityQueue<T extends Equatable> {
    private heap: HeapNode<T>[] = [];

    private swap(index1: number, index2: number) {
        const tmp = this.heap[index1];
        this.heap[index1] = this.heap[index2];
        this.heap[index2] = tmp;
    }

    private hasLeftChild(index: number) {
        return getLeftChildIndex(index) < this.heap.length;
    }

    private hasRightChild(index: number) {
        return getRightChildIndex(index) < this.heap.length;
    }

    private getPriority(index: number) {
        return this.heap[index].priority;
    }

    private bubbleUp(index: number) {
        let currentIndex = index;

        while(hasParent(currentIndex) && this.getPriority(currentIndex) < this.getPriority(getParent(currentIndex))) {
            const parent = getParent(currentIndex);
            this.swap(currentIndex, parent);
            currentIndex = parent;
        }
    }

    private bubbleDown(index: number) {
        let currentIndex = index;

        while(this.hasLeftChild(currentIndex)) {
            let smallerChildIndex = getLeftChildIndex(currentIndex);

            if(this.hasRightChild(currentIndex) && this.getPriority(getRightChildIndex(currentIndex)) < this.getPriority(getLeftChildIndex(currentIndex))) {
                smallerChildIndex = getRightChildIndex(currentIndex);
            }

            if(this.heap[currentIndex].priority < this.heap[smallerChildIndex].priority) {
                break;
            }

            this.swap(currentIndex, smallerChildIndex);
            currentIndex = smallerChildIndex;
        }
    }

    add(item: T, priority: number): void {
        let index = this.heap.length;
        this.heap.push({ value: item, priority });

        this.bubbleUp(index);
    }

    peek(): T {
        if(this.isEmpty()) throw new Error('Empty heap');

        return this.heap[0].value;
    }

    pop(): T {
        if(this.isEmpty()) throw new Error('Empty heap');

        this.swap(0, this.heap.length - 1);

        const { value } = this.heap.pop()!;

        this.bubbleDown(0);

        return value;
    }

    removeIndex(index: number) {
        if(index === this.heap.length - 1) {
            this.heap.pop();
            return;
        }

        this.swap(index, this.heap.length - 1);
        this.heap.pop();


        if(hasParent(index) && this.getPriority(index) < this.getPriority(getParent(index))) {
            this.bubbleUp(index);
        } else {
            this.bubbleDown(index);
        }
    }

    remove(item: T) {
        const index = this.heap.findIndex(heapNode => heapNode.value.equals(item));

        if(index === -1) throw new Error('Cannot find item');

        this.removeIndex(index);
    }

    isEmpty(): boolean {
        return this.heap.length === 0;
    }
}