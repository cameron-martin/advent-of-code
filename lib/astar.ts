import { HashTable, Hashable } from "./hashtable";
import { PriorityQueue } from "./priority-queue";
import { HashSet } from "./hashset";

export interface Child<T> {
    node: T;
    weight: number;
}

function reconstruct_path<T extends Hashable>(cameFrom: HashTable<T, T>, current: T) {
    const path = [current];
    while(cameFrom.has(current)) {
        current = cameFrom.get(current)!;
        path.push(current);
    }

    return path;
}

export function aStar<T extends Hashable>(start: T, goal: T, heuristic: (a: T, b: T) => number, children: (node: T) => Child<T>[]) {
    const closedSet = new HashSet<T>();
    const openQueue = new PriorityQueue<T>();
    const openSet = new HashSet<T>();
    openQueue.add(start, heuristic(start, goal));
    openSet.add(start);

    const cameFrom = new HashTable<T, T>();

    const gScore = new HashTable<T, number>();
    gScore.set(start, 0);

    const fScore = new HashTable<T, number>();
    fScore.set(start, heuristic(start, goal));

    let current;
    while(!openQueue.isEmpty()) {
        current = openQueue.pop();
        openSet.remove(current);

        if(current.equals(goal)) {
            return { path: reconstruct_path(cameFrom, current), distance: gScore.get(current)! };
        }

        closedSet.add(current);

        for(const neighbour of children(current)) {
            if(closedSet.has(neighbour.node)) continue;

            const tentative_gScore = gScore.get(current)! + neighbour.weight;

            if (!openSet.has(neighbour.node)) {
                openSet.add(neighbour.node);
                openQueue.add(neighbour.node, tentative_gScore + heuristic(neighbour.node, goal))
            } else if(tentative_gScore >= gScore.get(neighbour.node)!) {
                continue;
            }

            cameFrom.set(neighbour.node, current);
            gScore.set(neighbour.node, tentative_gScore);
            const newfScore = tentative_gScore + heuristic(neighbour.node, goal);
            if(newfScore !== fScore.get(neighbour.node)) {
                openQueue.remove(neighbour.node);
                openQueue.add(neighbour.node, newfScore);
                fScore.set(neighbour.node, newfScore);
            }
        }
    }

    throw new Error('Could not find shortest path');
}