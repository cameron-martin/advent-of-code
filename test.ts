import fs from 'fs-extra';
import path from 'path';
import { getPuzzleResult } from './lib/puzzle';

const solutionsPath = path.join(__dirname, 'solutions')
const years = fs.readdirSync(solutionsPath);

years.forEach(year => {
    describe(`Year ${year}`, () => {
        const yearPath = path.join(solutionsPath, year);
        const days = fs.readdirSync(yearPath);

        days.forEach(day => {            
            test(`Day ${day}`, async () => {
                const result = await getPuzzleResult(Number.parseInt(year), Number.parseInt(day));

                expect(result).toMatchSnapshot();
            });
        });

    });
});