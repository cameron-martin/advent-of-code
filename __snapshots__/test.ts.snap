// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Year 2015 Day 1 1`] = `
Object {
  "part1": 74,
  "part2": 1795,
}
`;

exports[`Year 2015 Day 25 1`] = `
Object {
  "part1": 19980801,
}
`;

exports[`Year 2018 Day 1 1`] = `
Object {
  "part1": 595,
  "part2": 80598,
}
`;

exports[`Year 2018 Day 2 1`] = `
Object {
  "part1": 6888,
  "part2": "icxjvbrobtunlelzpdmfkahgs",
}
`;

exports[`Year 2018 Day 3 1`] = `
Object {
  "part1": 104241,
  "part2": 806,
}
`;

exports[`Year 2018 Day 4 1`] = `
Object {
  "part1": 19025,
  "part2": 23776,
}
`;

exports[`Year 2018 Day 5 1`] = `
Object {
  "part1": 11298,
  "part2": 5148,
}
`;

exports[`Year 2018 Day 6 1`] = `
Object {
  "part1": 4398,
  "part2": 39560,
}
`;

exports[`Year 2018 Day 7 1`] = `
Object {
  "part1": "GKCNPTVHIRYDUJMSXFBQLOAEWZ",
  "part2": 1265,
}
`;

exports[`Year 2018 Day 8 1`] = `
Object {
  "part1": 40908,
  "part2": 25910,
}
`;

exports[`Year 2018 Day 9 1`] = `
Object {
  "part1": 373597,
  "part2": 2954067253,
}
`;

exports[`Year 2018 Day 22 1`] = `
Object {
  "part1": 10115,
  "part2": 990,
}
`;

exports[`Year 2018 Day 23 1`] = `
Object {
  "part1": 410,
  "part2": "",
}
`;

exports[`Year 2020 Day 1 1`] = `
Object {
  "part1": 719796,
  "part2": 144554112,
}
`;

exports[`Year 2020 Day 2 1`] = `
Object {
  "part1": 506,
  "part2": 443,
}
`;
